### PR TITLE
Initial review

### PR DESCRIPTION
**Pull request by edeandrea**
Review of demo/workshop

### CODE REVIEW

**Comment by edeandrea** [comment]
Couple of notes:
1. I upgraded it to Quarkus 3.6.0. Anything after that and the `OrderServiceTests.testEventIsPublishedWhenOrderIsCreated` fails. There's some race condition in there.
2. In `OrderResource` I removed the resteasy reactive specific classes in favor of the standard JAX-RS classes. In general its probably best to stick to the standards until/unless something above & beyond is needed
3. I renamed `BaseIntegrationTest` to `BaseTest`, since Quarkus considers these unit tests and not integration tests.
4. I added a `configKey` to the `PastryAPIClient` so that its configuration would be much simpler
5. I removed all instances where a `private static final Logger` is created and instead replaced with the Quarkus `Log` static method call

One thing I noticed is that the imports got all shifted around in any class that I touched. Probably due to my IDE settings (which I turned off for the project, but nonetheless they still got messed up). Sorry about that! If its that important, maybe we add the `impsort` plugin to the project?

Another thought I had was to make `OrderService.placeOrder` a reactive method that returns `Uni<Order>`. Doing so would remove the need for a lot of the `CompletableFuture` and joining that you are doing in that method. Combining multiple Unis into one is pretty trivial.

**Comment by edeandrea** [comment]
If you want, I can eliminate the `CompletableFuture` and replace with `Uni` without having to make the whole thing reactive...

```java
package org.acme.order.service;

import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;

import org.acme.order.client.PastryAPIClient;
import org.acme.order.service.model.Order;
import org.acme.order.service.model.OrderEvent;
import org.acme.order.service.model.OrderInfo;
import org.acme.order.service.model.ProductQuantity;
import org.eclipse.microprofile.rest.client.inject.RestClient;

import io.quarkus.logging.Log;

import io.smallrye.mutiny.Uni;
import io.smallrye.mutiny.infrastructure.Infrastructure;

/**
 * OrderService is responsible for checking business rules/constraints on Orders.
 * @author laurent
 */
@ApplicationScoped
public class OrderService {
   private record PastryAvailability(String pastryName, boolean available) {}

   // This is a dumb implementation of an event sourcing repository. Don't use this in production!
   private final Map<String, List<OrderEvent>> orderEventsRepository = new HashMap<>();

   @Inject
   @RestClient
   PastryAPIClient pastryRepository;

   @Inject
   OrderEventPublisher eventPublisher;

   /**
    * This method will check that an Order can be actually placed and persisted. A full implementation
    * will probably check stocks, customer loyalty, payment methods, shipping details, etc... For sake
    * of simplicity, we'll just check that products (here pastries) are all available.
    * @param info The order information.
    * @return A created Order with incoming info, new unique identifier and created status.
    * @throws UnavailablePastryException
    * @throws Exception
    */
   public Order placeOrder(OrderInfo info) throws UnavailablePastryException, Exception {
      // For all products in order, check the availability calling the Pastry API.
      var availabilityUnis = info.productQuantities().stream()
          .map(ProductQuantity::productName)
          .map(this::checkPastryAvailability)
          .toList();

      // Wait for all Unis to finish.
      var unavailablePastry = Uni.join()
          .all(availabilityUnis)
          .andFailFast()
          .runSubscriptionOn(Infrastructure.getDefaultWorkerPool())
          .await().atMost(Duration.ofSeconds(10))
          .stream()
          .filter(availability -> !availability.available())
          .findAny();

      if (unavailablePastry.isPresent()) {
         var pastryName = unavailablePastry.get().pastryName();
         throw new UnavailablePastryException(pastryName, "Pastry %s is not available".formatted(pastryName));
      }

      // Everything is available! Create (and probably persist ;-) a new order.
      Order result = new Order();
      result.setCustomerId(info.customerId());
      result.setProductQuantities(info.productQuantities());
      //result.setProductQuantities(List.of(new ProductQuantity("Millefeuille", 1)));
      result.setTotalPrice(info.totalPrice());

      // Persist and publish creation event.
      OrderEvent orderCreated = new OrderEvent(System.currentTimeMillis(), result, "Creation");
      persistOrderEvent(orderCreated);
      eventPublisher.publishOrderCreated(orderCreated);

      return result;
   }

   /**
    *
    * @param reviewedOrderEvent
    */
   public void updateReviewedOrder(OrderEvent reviewedOrderEvent) {
      persistOrderEvent(reviewedOrderEvent);
   }

   /**
    *
    * @param id
    * @return
    */
   public Order getOrder(String id) throws OrderNotFoundException {
      List<OrderEvent> orderEvents = orderEventsRepository.get(id);
      if (orderEvents != null) {
         return orderEvents.get(orderEvents.size() - 1).order();
      }
      throw new OrderNotFoundException(id);
   }

   /**
    *
    * @param id
    * @return
    * @throws OrderNotFoundException
    */
   public List<OrderEvent> getOrderEvents(String id) throws OrderNotFoundException {
      List<OrderEvent> orderEvents = orderEventsRepository.get(id);
      if (orderEvents == null) {
         throw new OrderNotFoundException(id);
      }
      return orderEvents;
   }

   private Uni<PastryAvailability> checkPastryAvailability(String pastryName) {
      return Uni.createFrom().item(() -> {
             Log.infof("Checking pastry availability for pastry %s", pastryName);
             return new PastryAvailability(pastryName, "available".equals(pastryRepository.getPastry(pastryName).status()));
          })
          .emitOn(Infrastructure.getDefaultWorkerPool())
          .onFailure().invoke(e -> Log.errorf(e, "Got exception from Pastry client: %s", e.getMessage()))
          .onFailure().recoverWithItem(new PastryAvailability(pastryName, false));
   }

   private void persistOrderEvent(OrderEvent event) {
      List<OrderEvent> orderEvents = orderEventsRepository.get(event.order().getId());
      if (orderEvents == null) {
         orderEvents = new ArrayList<>();
      }
      orderEvents.add(event);
      orderEventsRepository.put(event.order().getId(), orderEvents);
   }
}
```

or I can just clean up a bit...

```java
package org.acme.order.service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;

import org.acme.order.client.PastryAPIClient;
import org.acme.order.service.model.Order;
import org.acme.order.service.model.OrderEvent;
import org.acme.order.service.model.OrderInfo;
import org.acme.order.service.model.ProductQuantity;
import org.eclipse.microprofile.rest.client.inject.RestClient;

import io.quarkus.logging.Log;

/**
 * OrderService is responsible for checking business rules/constraints on Orders.
 * @author laurent
 */
@ApplicationScoped
public class OrderService {
   private record PastryAvailability(String pastryName, boolean available) {}

   // This is a dumb implementation of an event sourcing repository. Don't use this in production!
   private final Map<String, List<OrderEvent>> orderEventsRepository = new HashMap<>();

   @Inject
   @RestClient
   PastryAPIClient pastryRepository;

   @Inject
   OrderEventPublisher eventPublisher;

   /**
    * This method will check that an Order can be actually placed and persisted. A full implementation
    * will probably check stocks, customer loyalty, payment methods, shipping details, etc... For sake
    * of simplicity, we'll just check that products (here pastries) are all available.
    * @param info The order information.
    * @return A created Order with incoming info, new unique identifier and created status.
    * @throws UnavailablePastryException
    * @throws Exception
    */
   public Order placeOrder(OrderInfo info) throws UnavailablePastryException, Exception {
      // For all products in order, check the availability calling the Pastry API.
      var availabilityFutures = info.productQuantities().stream()
          .map(ProductQuantity::productName)
          .map(this::checkPastryAvailability)
          .toList();

      // Wait for all completable futures to finish.
      CompletableFuture.allOf(availabilityFutures.toArray(new CompletableFuture[availabilityFutures.size()])).join();

      var unavailablePastry = availabilityFutures.stream()
          .map(availabilityFuture -> {
              try {
                  return availabilityFuture.get();
              }
              catch (InterruptedException | ExecutionException e) {
                  throw new RuntimeException("Unexpected exception: " + e.getMessage());
              }
          })
          .filter(availability -> !availability.available())
          .findAny();

      if (unavailablePastry.isPresent()) {
         var pastryName = unavailablePastry.get().pastryName();
         throw new UnavailablePastryException(pastryName, "Pastry %s is not available".formatted(pastryName));
      }

      // Everything is available! Create (and probably persist ;-) a new order.
      Order result = new Order();
      result.setCustomerId(info.customerId());
      result.setProductQuantities(info.productQuantities());
      //result.setProductQuantities(List.of(new ProductQuantity("Millefeuille", 1)));
      result.setTotalPrice(info.totalPrice());

      // Persist and publish creation event.
      OrderEvent orderCreated = new OrderEvent(System.currentTimeMillis(), result, "Creation");
      persistOrderEvent(orderCreated);
      eventPublisher.publishOrderCreated(orderCreated);

      return result;
   }

   /**
    *
    * @param reviewedOrderEvent
    */
   public void updateReviewedOrder(OrderEvent reviewedOrderEvent) {
      persistOrderEvent(reviewedOrderEvent);
   }

   /**
    *
    * @param id
    * @return
    */
   public Order getOrder(String id) throws OrderNotFoundException {
      List<OrderEvent> orderEvents = orderEventsRepository.get(id);
      if (orderEvents != null) {
         return orderEvents.get(orderEvents.size() - 1).order();
      }
      throw new OrderNotFoundException(id);
   }

   /**
    *
    * @param id
    * @return
    * @throws OrderNotFoundException
    */
   public List<OrderEvent> getOrderEvents(String id) throws OrderNotFoundException {
      List<OrderEvent> orderEvents = orderEventsRepository.get(id);
      if (orderEvents == null) {
         throw new OrderNotFoundException(id);
      }
      return orderEvents;
   }

   private CompletableFuture<PastryAvailability> checkPastryAvailability(String pastryName) {
      return CompletableFuture.supplyAsync(() -> {
             Log.infof("Checking pastry availability for pastry %s", pastryName);
             return new PastryAvailability(pastryName, "available".equals(pastryRepository.getPastry(pastryName).status()));
          })
          .exceptionally(e -> {
             Log.errorf(e, "Got exception from Pastry client: %s", e.getMessage());
             return new PastryAvailability(pastryName, false);
          });
   }

   private void persistOrderEvent(OrderEvent event) {
      List<OrderEvent> orderEvents = orderEventsRepository.get(event.order().getId());
      if (orderEvents == null) {
         orderEvents = new ArrayList<>();
      }
      orderEvents.add(event);
      orderEventsRepository.put(event.order().getId(), orderEvents);
   }
}
```

The `Uni` way eliminates the ugly try / catch inside a lambda, since it will handle & propagate the failure when calling `.await()`.

File: step-1-getting-started.md [hunk 1]
```diff
@@ -11,30 +11,30 @@ Testcontainers libraries are compatible with Java 8+, but this workshop uses a Q
 
 We would recommend using [SDKMAN](https://sdkman.io/) to install Java on your machine if you are using MacOS, Linux or Windows WSL.
 
-### Install Docker
+### Install a container runtime
 
-You need to have a Docker environment to use Testcontainers.
+You need to have a [Docker](https://docs.docker.com/get-docker/) or [Podman](https://podman.io/) environment to use Testcontainers.
```
Comment (lbroudoux) [inline step-1-getting-started.md:16]: I heard that Podman with Testcontainers was a bit clunky, but I never tested it though... Do you have experience with that?

**Review by lbroudoux — APPROVED** [review]
Wouah! This is what a called a review and big improvements. Thnak you so much Eric for the time you spent on this! This make the demo way more adapted to Quarkus idioms.

**Comment by lbroudoux** [comment]
Thanks again for the review and the massive improvement on the Async/CompletableFuture stuff. The initial intent was to make parallel calls, but I couldn't make it work quickly and switched to another topic ;-) 
I admit that I prefer the `CompletableFuture` version as I'm not accustomed to Mutiny stuff... but if you think it's way more representative of the way people code in Quarkus app, then we can go for this evolution!

**Comment by edeandrea** [comment]
We can certainly do the `CompletableFuture` way. It isn't wrong. It's plain old java. Just having a try/catch inside a lambda expression is clunky, but I can probably create a private method for that logic, then it just becomes a function reference. This will also clean things up from a map keyed by the future to just a list. I created a private record in the class which holds the result of the future combined with the name of the pastry it is referencing.

I'll send another PR with that change a bit later.

File: step-1-getting-started.md (hunk 1 above)
Comment (edeandrea) [reply step-1-getting-started.md:16]: We have a whole guide dedicated to it (https://quarkus.io/guides/podman). I'll add that to the docs in another PR.

**Pull request merged**